### PR TITLE
Feat/flow 8078 savencia

### PR DESCRIPTION
**Pull request by lmancel**

### CODE REVIEW

File: webpack.main.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
+const { getOptimization, webpackModule, resolve, Mode, output, deleteDir, OUTPUT_DIR } = require('./webpack/utils')
+
+module.exports = async (env) => {
+    await deleteDir(OUTPUT_DIR)
```
Comment (lmancel) [inline webpack.main.js]: Will replace with CleanWebpackPlugin

File: webpack.config.js [hunk 2]
```diff
@@ -0,0 +1,249 @@
+
```
Comment (lmancel) [inline webpack.config.js]: Will delete this file

File: static/pages/config.html [hunk 3]
```diff
@@ -105,19 +117,24 @@ <h3>Debug</h3>
...
-		externalUrl = document.getElementById('extUrl').value
-		ipc.send('setExtUrl',externalUrl)
+	function validateExtUrl() {
+		const url = document.getElementById('extUrl').value
+		try {
+			// try to parse as URL to check validity
+			new URL(url)
+			ipc.send('setExtUrl', url)
+		} catch (e) {
+			invalidUrl()
```
Comment (lmancel) [inline static/pages/config.html:133]: Will add `console.error` for debugging

File: src/wexond/main/view.ts [hunk 4]
```diff
@@ -1,14 +1,198 @@
...
+  // BrowserViewConstructorOptions,
+} from 'electron'
+import { appWindow } from '.'
+import { sendToAllExtensions } from './extensions'
+import { engine } from './services/web-request'
+import { settings } from './index'
+import { parse } from 'tldts'
+import { buildPreloadPath } from '../../common/preload'
+
+// function BrowserWindowExtend(electronClass: BrowserWindow) {
```
Comment (lmancel) [inline src/wexond/main/view.ts]: This was wip, I will remove it

File: src/application-manager/application-manager.ts [hunk 5]
```diff
@@ -119,16 +94,16 @@ export class ApplicationManager {
 
   initLocalApps(clearStore: boolean = false): Promise<void[]> {
     return new Promise((resolve, reject) => {
-      fs.readdir(join(app.getAppPath(), 'build', 'applications'), (err, files) => {
+      // TODO: open ticket @ electron "withFileTypes" option is not working inside of the asar archive, it only returns the file names
```
Comment (ArnaudGathy) [inline src/application-manager/application-manager.ts]: What about we create a ticket with that task 😊

File: src/applications/keyboard/keyboardWindow.ts [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+      ...buildPositionFromParents(parent.getBounds()),
+    })
+    const url = buildFileUrl('keyboard')
+    this.loadURL(url)
+
+    const ipcEvents = {
+      keyPress: this.onKeyPress.bind(this),
+      keyUp: this.onKeyUp.bind(this),
+    }
+    Object.entries(ipcEvents).forEach(event => ipcMain.on(...event))
```
Comment (ArnaudGathy) [inline src/applications/keyboard/keyboardWindow.ts:47]: Neat syntax 👌

File: src/applications/keyboard/preload.ts [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+}
+
+const ipcRenderer = require('electron').ipcRenderer
+
+process.once('loaded', () => {
+  window.global = window.global || (window as any)
+  global.ipcRenderer = ipcRenderer
+})
+
+export {}
```
Comment (ArnaudGathy) [inline src/applications/keyboard/preload.ts]: Why ? What does it do ?

File: src/applications/keyboard/views/style.css [hunk 8]
```diff
@@ -0,0 +1,3 @@
+.hg-button.hg-highlight {
+    box-shadow: 0 0 40px 5px rgba(64, 150, 255, 0.696);
```
Comment (ArnaudGathy) [inline src/applications/keyboard/views/style.css:2]: That's an oddly specific value for opacity 🤔

File: src/barcoKeyboard/barcoKeyBoardController.ts [hunk 9]
```diff
@@ -1,15 +1,53 @@
...
 
-class BarcoKeyBoardController implements BarcoKeyBoardControllerInterface {
-  async open(): Promise<void> {
-    await fetch('http://localhost:9000/keyboard/open', { method: 'GET', mode: 'no-cors' })
+class Keyboard {
+  flowrStore?: Store<IFlowrStore>
+  keyboardWindow?: KeyboardWindow
+
+  get isEnabled(): boolean {
+    return this.flowrStore?.get('enableVirtualKeyboard') ?? false
```
Comment (ArnaudGathy) [inline src/barcoKeyboard/barcoKeyBoardController.ts]:
```suggestion
    return !!this.flowrStore?.get('enableVirtualKeyboard')
```

File: src/barcoKeyboard/keyboardMixin.ts [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+        },
+        [VirtualKeyboardEvent.TOGGLE]: (e: IpcMainEvent) => {
+          if (e.sender === this.webContents && keyboard.isEnabled) {
+            keyboard.toggle(this)
+          }
+        },
+      }
+
+      Object.entries(ipcEvents).forEach(event => ipcMain.on(...event))
+      this.on('close', () => Object.entries(ipcEvents).forEach(event => ipcMain.removeListener(...event)))
```
Comment (ArnaudGathy) [inline src/barcoKeyboard/keyboardMixin.ts]: There are the same two lines in `keyboardWindow.ts`

File: src/barcoKeyboard/keyboardOpener.ts [hunk 11]
```diff
@@ -0,0 +1,21 @@
+import { IpcRenderer } from 'electron'
+import { VirtualKeyboardEvent } from './events'
+
+// tslint:disable-next-line: function-name
+export function KeyboardOpener<T extends new (...args: any[]) => {}>(yo: T) {
```
Comment (ArnaudGathy) [inline src/barcoKeyboard/keyboardOpener.ts]: Is it missing a `=` to make an arrow function `=>` or is it some syntax i've never seen ?
`{}>(yo: T)`

File: src/barcoKeyboard/keyboardOpener.ts [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+  return class extends yo {
```
Comment (ArnaudGathy) [inline src/barcoKeyboard/keyboardOpener.ts]: What is `yo` ?

File: src/wexond/main/view.ts (hunk 4 above)
Comment (ArnaudGathy) [inline src/wexond/main/view.ts]: Is this big chunk of commented code needed ?

File: src/wexond/renderer/app/components/App/index.tsx [hunk 13]
```diff
@@ -3,26 +3,27 @@ import * as React from 'react';
...
+import { App } from '../../models/app'
 
 const GlobalStyle = createGlobalStyle`${Style}`;
 
 window.onbeforeunload = () => {
   ipcRenderer.send('browserview-clear');
 };
 
-export const App = observer(() => {
+export default observer((data: App) => {
```
Comment (ArnaudGathy) [inline src/wexond/renderer/app/components/App/index.tsx:21]: Why this change ? Just to not have the same name as `App` in `models/app` ?

File: src/wexond/renderer/app/components/Toolbar/index.tsx [hunk 14]
```diff
@@ -53,46 +52,57 @@ export const Toolbar = observer(() => {
...
+          {buttons}
+        </StyledToolbar>
+    )
   }
 
   return (
       <StyledToolbar isHTMLFullscreen={store.isHTMLFullscreen}>
         <NavigationButtons />
         <Tabbar />
         <Find />
```
Comment (ArnaudGathy) [inline src/wexond/renderer/app/components/Toolbar/index.tsx]: The only difference between the two renders are `<Tabbar />` and `<Find />`
It would have been much easier and less awkward to condtionnaly render them instead of having two returns and all the buttons stored in a variable

```
{!data.disableTabs && (<>
    <Tabbar />
    <Find />
</>)}
```

File: src/barcoKeyboard/keyboardOpener.ts (hunk 11 above)
Comment (lmancel) [reply src/barcoKeyboard/keyboardOpener.ts]: I'll delete this file, it was an attempt that was unneeded in the end

File: src/applications/keyboard/preload.ts (hunk 7 above)
Comment (lmancel) [reply src/applications/keyboard/preload.ts]: Hmmmmm right... I think it was to be able to use a `declare` statement above so it's considered as a module,  but since there already is an import statement I might be able to remove it

File: src/applications/keyboard/views/style.css (hunk 8 above)
Comment (lmancel) [reply src/applications/keyboard/views/style.css:2]: Just used the color selector from VSCode, it's more or less random 😄

File: src/wexond/main/view.ts (hunk 4 above)
Comment (lmancel) [reply src/wexond/main/view.ts]: No it's not, I did not push the message I wrote earlier but it's going to be removed.

It's related to the fact that extending electron's base classes is not supported, so I wanted to create a simple class that would implement the exact same interface and just act as a bridge. Might be needed in the future but overkill for now.

File: src/wexond/renderer/app/components/App/index.tsx (hunk 13 above)
Comment (lmancel) [reply src/wexond/renderer/app/components/App/index.tsx:21]: Actually I just used the same logic than what was done in other Wexond components without even trying to think of something better...

File: src/applications/keyboard/keyboardWindow.ts [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+  return {
+    width: w,
+    height: h,
+    x: Math.round(.2 * width),
+    y: Math.round(parentRectangle.y + parentRectangle.height - h),
+  }
+}
+
+export class KeyboardWindow extends BrowserWindow {
+  capsLock: boolean = false
```
Comment (Azevedo-Dani) [inline src/applications/keyboard/keyboardWindow.ts]:
```suggestion
  capsLock = false
```

File: src/applications/keyboard/views/index.ts [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+        return [{
+          class: 'hg-highlight',
+          buttons: SimpleKeyboardKeys.SHIFT,
+        }]
+      case KeyboardState.CAPS_LOCK:
+        return [{
+          class: 'hg-highlight',
+          buttons: SimpleKeyboardKeys.CAPS_LOCK,
+        }]
+      case KeyboardState.DEFAULT:
```
Comment (Azevedo-Dani) [inline src/applications/keyboard/views/index.ts]:
```suggestion
```

File: src/applications/keyboard/views/index.ts [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+        }
+      case SimpleKeyboardKeys.CAPS_LOCK:
+        switch (this.state) {
+          case KeyboardState.CAPS_LOCK:
+            return KeyboardState.DEFAULT
+          default:
+            return KeyboardState.CAPS_LOCK
+        }
+      default:
+        switch (this.state) {
```
Comment (Azevedo-Dani) [inline src/applications/keyboard/views/index.ts]: We don't have a default case for this switch ?

File: src/inactivity/window.ts [hunk 18]
```diff
@@ -0,0 +1,42 @@
+import { BrowserView, BrowserWindow, IpcMainEvent } from 'electron'
+import { throttle } from 'lodash'
+import { ACTIVITY_EVENT, INACTIVITY_THROTTLE } from './utils'
+
+export async function watchForInactivity(baseWindow: BrowserWindow | BrowserView, timeoutDuration: number, callback: (baseWindow: BrowserWindow | BrowserView) => any) {
+  let timeout: number | undefined = undefined
```
Comment (Azevedo-Dani) [inline src/inactivity/window.ts:6]:
```suggestion
  let timeout?: number
```

Comment (lmancel) [reply src/inactivity/window.ts:6]: I can't use this syntax, it's not a class property
Screenshot 2020-11-16 at 10 47 16

File: src/applications/keyboard/views/index.ts (hunk 17 above)
Comment (lmancel) [reply src/applications/keyboard/views/index.ts]: The default case is supposed to do nothing. It could be replaced with an `if` statement, I just kept the same syntax as for the rest of the function

File: src/application-manager/application-manager.ts (hunk 5 above)
Comment (lmancel) [reply src/application-manager/application-manager.ts]: Actually already done by other peoples: https://github.com/electron/electron/issues/25349

Also cf https://github.com/electron/electron/pull/24062

**Review by Azevedo-Dani — APPROVED** [review]

**Review by ArnaudGathy — APPROVED** [review]
